staging/src/k8s.io/endpointslice/topologies/manager.go

```diff
@@ -0,0 +1,240 @@
+		return nil
+	}
+
+	events := []*EventBuilder{{
+		EventType: corev1.EventTypeWarning,
+		Reason:    "TopologyAwareHintsDisabled",
+		Message:   FormatWithAddressTypeAndTopologyName(TopologyAwareHintsDisabled, addrType, activeTopology.Name()),
+	}}
+
+	m.setActiveTopologyForService(serviceKey, addrType, nil)
+	activeTopology.ClearCachedHints(logger, serviceKey, addrType)
+	return events
+}
+
+// activeTopologyForService returns the Topology which is presently being used
+// by the service.
+//
+// The Topology returned represents the "current state", which could be
+// different from the "desired state". Use desiredTopologyForService() to get
+// the "desired state".
+func (m *Manager) activeTopologyForService(serviceKey string, addrType discoveryv1.AddressType) (Topology, bool) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.activeTopologiesByService[serviceKey]; !ok {
+		m.activeTopologiesByService[serviceKey] = make(map[discoveryv1.AddressType]Topology)
+	}
+
+	topology, ok := m.activeTopologiesByService[serviceKey][addrType]
+	return topology, ok && topology != nil
+}
+
+// setActiveTopologyForService sets the active Topology for the service and
+// addrType.
+//
+// Refer comment under activeTopologyForService() for an understanding of what
+// active Topology means.
+func (m *Manager) setActiveTopologyForService(serviceKey string, addrType discoveryv1.AddressType, topology Topology) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if _, ok := m.activeTopologiesByService[serviceKey]; !ok {
+		m.activeTopologiesByService[serviceKey] = make(map[discoveryv1.AddressType]Topology)
+	}
+
+	m.activeTopologiesByService[serviceKey][addrType] = topology
+}
+
+// desiredTopologyForService returns the Topology corresponding to the topology
+// annotation in the service.
+//   - enabled return value signifies that the topology annotation determines an
+//     enabled value.
+//   - exists return value signifies existence of the topology in the topology
+//     manager.
+//
+// The Topology returned represents the "desired state", which could be
+// different from the "current state". Use activeTopologyForService() to get
+// the "current state".
+func (m *Manager) desiredTopologyForService(service *corev1.Service) (topology Topology, enabled, exists bool) {
+	desiredTopologyName, desiredTopologyEnabled := TopologyNameFromAnnotations(service.GetAnnotations())
+	if !desiredTopologyEnabled {
+		return nil, false, false
+	}
+
+	// Special case for backward compatibility: If the topology annotation value
+	// is set to "Auto", we need to return defaultTopology.
+	if desiredTopologyName == "Auto" || desiredTopologyName == "auto" {
+		topology, ok := m.topologiesByName[m.defaultTopologyName]
+		return topology, true, ok
+	}
+
+	for topologyName, topology := range m.topologiesByName {
+		if desiredTopologyName == topologyName || desiredTopologyName == strings.ToLower(topologyName) { //nolint:gocritic // strings.EqualFold() does not exactly match the intent.
```

Comment:
I think it would be fair to make this a case-sensitive check. The support for "auto" and "Auto" is leftover from an unfortunate mistake on my part where we started with support for "auto" before realizing that "Auto" is more consistent with k8s conventions. We left in support for "auto" for the sake of backwards compatibility. In this case we're talking about entirely new values, so it's probably best to make them case-sensitive matches from the outset.